server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -460,8 +509,37 @@ private ThreadContextStruct putResponse(final String key, final String value, fi
             } else {
                 newResponseHeaders.put(key, Collections.singletonList(value));
             }
+            return new ThreadContextStruct(requestHeaders, newResponseHeaders, transientHeaders,
+                isSystemContext, wrnHeadersSize + curWrnHeaderSize, isWrnLmtReached);
+        }
 
-            return new ThreadContextStruct(requestHeaders, newResponseHeaders, transientHeaders, isSystemContext);
+        //replace last warning header(s) with "headers limit reached" warning
+        //respecting limitations on headers size if it is set by user
+        private ThreadContextStruct addWrnLmtReachedHeader(){
```

Comment:
The user already expects if they have set this setting that warning headers will be truncated. Therefore, I don't think we should do this, losing a warning header (or more) at the expense of a warning header saying that there are more warnings that can be found in the logs. Instead I think that we should log (in the main Elasticsearch log) when we truncate.